src/main/scala/TableGroup.scala

```diff
@@ -1,6 +1,412 @@
+          tableUrl,
+          i,
+          reference,
+          resourceNode
+        )
+        val tableColumns: mutable.Map[String, Column] = Map()
+        for (column <- referencedTable.columns) {
+          column.name
+            .foreach(columnName => tableColumns += (columnName -> column))
+        }
+
+        val referencedColumns: Array[Column] = reference
+          .get("columnReference")
+          .elements()
+          .asScalaArray
+          .map(columnReference => {
+            tableColumns.get(columnReference.asText()) match {
+              case Some(column) => column
+              case None =>
+                throw new MetadataError(
+                  s"column named ${columnReference.asText()} does not exist in #{resource}," +
+                    s" $$.tables[?(@.url = '${tableUrl}')].tableSchema.foreign_keys[${i}].reference.columnReference"
+                )
+            }
+          })
+
+        val foreignKeyWithTable =
+          ForeignKeyWithTable(foreignKey, referencedTable, referencedColumns)
+        referencedTable.foreignKeyReferences :+= foreignKeyWithTable
+      }
+    }
+  }
+
+  private def setReferencedTableOrThrowException(
+      baseUrl: String,
+      tables: Map[String, Table],
+      tableUrl: String,
+      i: Int,
+      reference: JsonNode,
+      resourceNode: JsonNode
+  ): Table = {
+    if (resourceNode.isMissingNode) {
+      val schemaReferenceNode = reference.get("schemaReference")
+      val schemaUrl =
+        new URL(new URL(baseUrl), schemaReferenceNode.asText()).toString
+      val referencedTables = List.from(
+        tables.values.filter(t =>
+          t.schemaId.isDefined && t.schemaId.get == schemaUrl
+        )
+      )
+      referencedTables match {
+        case referencedTable :: _ => referencedTable
+        case Nil =>
+          throw new MetadataError(
+            s"Could not find foreign key referenced schema ${schemaUrl}, " +
+              s"$$.tables[?(@.url = '${tableUrl}')].tableSchema.foreignKeys[${i}].reference.SchemaReference"
+          )
+      }
+    } else {
+      val resource = new URL(
+        new URL(baseUrl),
+        resourceNode.asText()
+      ).toString
+      tables.get(resource) match {
+        case Some(refTable) => refTable
+        case None =>
+          throw new MetadataError(
+            s"Could not find foreign key referenced table ${tableUrl}, " +
+              s"$$.tables[?(@.url = '${tableUrl}')].tableSchema.foreignKeys[${i}].reference.resource"
+          )
+      }
+    }
+  }
+
+  private def ensureTypeofTableGroup(tableGroupNode: ObjectNode): Unit = {
+    val typeNode = tableGroupNode.path("@type")
+    if (!typeNode.isMissingNode && typeNode.asText != "TableGroup") {
+      throw new MetadataError(
+        s"@type of table group is not 'TableGroup', found @type to be a '${typeNode.asText()}'"
+      )
+    }
+  }
+
+  def classifyPropertiesBasedOnPropertyTypeAndSetWarnings(
+      json: ObjectNode,
```

Comment:
Again, something bettern than `json` would be good here.

Reply:
Done